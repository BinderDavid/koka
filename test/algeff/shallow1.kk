//-------------------------------------------------------------
//
//-------------------------------------------------------------

effect await<a> {
  await() : a
}

effect send<a> {
  send( x : a ) : ()
}

function down(prod : () -> <send<c>,await<c>|e> a, action : () -> <await<c>,send<c>|e> a) : <send<c>,await<c>|e> a {
  handle shallow(action) {
    return x -> x
    await()  -> up(resume){ prod() } 
  }
}

function up(cons : (c) -> <await<c>,send<c>|e> a, action : () -> <send<c>,await<c>|e> a) : <send<c>,await<c>|e> a {
  handle shallow(action) {
    return x -> x
    send(x)  -> down{ resume(()) }{ cons(x) } 
  }
}

function producer() {
  send(1)
  send(2)
  0
}

function consumer() {
  val x = await()
  if (x > 0) then await() + 1 else await() - 1
}

val unawait = handler {
  return x -> x
  await()  -> error("unresolved await")
}


val unsend = handler {
  return x -> x
  send(x)  -> error("unresolved send")
}

function main() {
  unsend{ unawait{ down( producer, consumer ) } } 
}