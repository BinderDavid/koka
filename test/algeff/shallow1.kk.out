3
.Eff-await : forall<a,b> (.ops-await<b,a>) -> operation<a>
.Eff-send : forall<a,b> (.ops-send<b,a>) -> operation<a>
.Op-await : forall<a,b> .ops-await<a,b>
.Op-send : forall<a,b> (a) -> .ops-send<a,b>
.copy : forall<a,b> (.ops-send<a,b>, a) -> .ops-send<a,b>
.copy : forall<a,b> (.ops-await<a,b>) -> .ops-await<a,b>
await : forall<a> () -> (await<a>) a
consumer : () -> (await<int>) int
down : forall<a,b,e> (prod : () -> <await<a>,send<a>,div|e> b, action : () -> <await<a>,send<a>,div|e> b) -> <await<a>,send<a>,div|e> b
main : () -> <pure,console> ()
producer : () -> (send<int>) int
send : forall<a> (x : a) -> (send<a>) ()
unawait : forall<a,b,e> (action : () -> <await<a>,exn|e> b) -> <exn|e> b
unsend : forall<a,b,e> (action : () -> <send<a>,exn|e> b) -> <exn|e> b
up : forall<a,b,e> (cons : (a) -> <await<a>,send<a>,div|e> b, action : () -> <await<a>,send<a>,div|e> b) -> <await<a>,send<a>,div|e> b